File: py-110/small_problems/medium_2/triangle_sides.py
"""
A triangle is classified as follows:

Equilateral: All three sides have the same length.
Isosceles: Two sides have the same length, while the third is different.
Scalene: All three sides have different lengths.

To be a valid triangle, the sum of the lengths of the two shortest sides must be greater than the length of the 
longest side, and every side must have a length greater than 0. If either of these conditions is not satisfied, 
the triangle is invalid.

Write a function that takes the lengths of the three sides of a triangle as arguments and returns one of the 
following four strings representing the triangle's classification: 'equilateral', 'isosceles', 'scalene', or 'invalid'.

P:
Determine whether the 3 given floats create an equilateral, isosceles, scalene or no triangle at all.

Rules:
    - A triangle is defined as the sum of the two shortest sides > longest side + every side > 0
    - Equilateral means all 3 sides have the same length
    - Isosceles means that 2 sides have the same length while the third is different
    - Scalene means that all 3 sides have different lengths
    - Assume we'll always receive 3 floats as input with value 0 or greater

E: Confirmed

Data structures:
    - Input: 3 floats, each representing the side of a triangle
    - Output: String, either "equilateral", "isosceles", "scalene", or "invalid"
    - Intermediary:
        - List / Tuple: List to store the lengths + use built-in methods to calculate the min or max sides
        - Boolean: Do the sides make up a valid triangle? Yes / no
        - Set: Determine the number of unique side "values" we got (1, 2, or 3 unique values)

High-level ideas:
    - Figure out whether the sides make a valid triangle first. If so, determine the type of triangle based on the number of
      unique side values given
        - To figure out whether the sides make a valid triangle, calculate the sum of the two shortest sides, check that 
          the sum is > than the longest side, and make sure every side > 0

A:
    - Store the inputs into a list 'sides'
    - (Helper) Figure out whether the sides make a valid triangle
        - If any one of the triangle sides is <= 0, return False
        - Find the two smallest sides and calculate their sum
        - If the sum is <= the longest side, return False
        - Otherwise, return True
    - If we have an invalid triangle, return "invalid"
    - Create a set using the input values
    - If the length of the set is 1, return "equilateral"
    - If the length is 2, return "isosceles"
    - Else, return "scalene"

"""

def is_triangle(sides):
    for side in sides:
        if side <= 0:
            return False

    sorted_sides = sorted(sides)
    *smallest, largest = sorted_sides
    if sum(smallest) <= largest:
        return False
    
    return True

def triangle(side1, side2, side3):
    sides = [side1, side2, side3]
    if not is_triangle(sides):
        return "invalid"
    
    unique_side_values = set(sides)
    num_unique = len(unique_side_values)
    if num_unique == 1:
        return "equilateral"
    elif num_unique == 2:
        return "isosceles"
    else:
        return "scalene"


# Test cases
print(triangle(3, 3, 3) == "equilateral")  # True
print(triangle(3, 3, 1.5) == "isosceles")  # True
print(triangle(3, 4, 5) == "scalene")      # True
print(triangle(0, 3, 3) == "invalid")      # True
print(triangle(3, 1, 1) == "invalid")      # True

"""
Time: 17.5 min
"""

File: ls-215/exercises/medium_2/triangle_sides.js
/*
A triangle is classified as follows:

Equilateral: All three sides are of equal length.
Isosceles: Two sides are of equal length, while the third is different.
Scalene: All three sides are of different lengths.

To be a valid triangle, the sum of the lengths of the two shortest sides must be greater than the length of the longest side, and every side must have a length greater than 0. If either of these conditions is not satisfied, the triangle is invalid.

Write a function that takes the lengths of the three sides of a triangle as arguments and returns one of the following four strings representing the triangle's classification: 'equilateral', 'isosceles', 'scalene', or 'invalid'.


My questions:
- Will we always get numbers as input? What happens if we get anything that isn't a number?
- What happens if we get fewer or greater than 3 inputs? What should our program do?
- Confirm that we might get a number < 0 as an input
- Will we always get integers? Are non-integer inputs still valid to work with?
- Do we have to worry about getting NaN values as input, or Infinity / -Infinity?
- Confirm that the order in which we receive inputs doesn't matter
- Confirm that if the sum of 2 shortest = longest, that still isn't a valid triangle


Input: 3 numbers
Output: String (one of 'equilateral', 'isosceles', 'scalene', or 'invalid')
Rules:
- Valid triangle = sum of 2 shortest > longest, all sides > 0
- Equilateral means all 3 sides are equal length
- Isosceles means 2 sides equal, 3rd is not
- Scalene means all 3 different
- Assume inputs always numbers
- Assume exactly 3 arguments will be provided
- We might receive negative inputs
- We might receive non-integer inputs as well
- Don't worry about NaNs or Infinity / -Infinity
- Order in which we get the inputs does not matter


Data structures:
- Sides -> array
- Count the unique side values -> object, or array


Algorithm:

High-level:
- Check if the numbers create a valid triangle
- If not, return "invalid"
- If so, figure out if triangle is equilateral, isosceles, or scalene and return the appropriate string

HELPER: isValidTriangle(sides) -> boolean
- Create a copy of input array -> `sidesCopy`
- Sort `sidesCopy` from smallest side to largest
- If 0th element of `sidesCopy` <= 0
  - Return false
- If sum of 0th and 1st elements are > 2nd element
  - Return true
- else
  - Return false

HELPER: determineTriangle(sides) -> string
- Create an empty array `unique`
- Iterate through `sides` -> `side`
  - If `side` is not in `unique`, add it
- If `unique` length is 1:
  - Return 'equilateral'
- Else if `unique` length is 2:
  - Return 'isosceles'
- Else
  - Return 'scalene'


Main:
- Put numbers into array -> `sides`
- Check if the numbers create a valid triangle
  - Use isValidTriangle
- If not, return "invalid"
- If so, figure out if triangle is equilateral, isosceles, or scalene and return the appropriate string
  - Use determineTriangle

*/

function isValidTriangle(sides) {
  let sidesCopy = [...sides];
  sidesCopy.sort((a, b) => a - b);

  let [ smallest, mid, largest ] = sidesCopy;
  if (smallest <= 0) {
    return false;
  }

  return smallest + mid > largest;
}

function determineTriangle(sides) {
  let unique = sides.reduce((acc, side) => {
    if (!acc.includes(side)) {
      acc.push(side);
    }
    return acc;
  }, []);

  if (unique.length === 1) {
    return 'equilateral';
  } else if (unique.length === 2) {
    return 'isosceles';
  } else {
    return 'scalene';
  }
}

function triangle(...sides) {
  if (!isValidTriangle(sides)) {
    return "invalid";
  }

  return determineTriangle(sides);
}


console.log(triangle(3, 3, 3));        // "equilateral"
console.log(triangle(3, 3, 1.5));      // "isosceles"
console.log(triangle(3, 4, 5));        // "scalene"
console.log(triangle(0, 3, 3));        // "invalid"
console.log(triangle(3, 1, 1));        // "invalid"

console.log(triangle(3, 1, 2));        // "invalid"
console.log(triangle(3, -1, 2));        // "invalid"
